nixos/modules/services/network-filesystems/samba.nix

```diff
@@ -3,35 +3,12 @@
 with lib;
 
 let
-
-  smbToString = x: if builtins.typeOf x == "bool"
-                   then boolToString x
-                   else toString x;
-
   cfg = config.services.samba;
 
-  samba = cfg.package;
-
-  shareConfig = name:
-    let share = getAttr name cfg.shares; in
-    "[${name}]\n " + (smbToString (
-       map
-         (key: "${key} = ${smbToString (getAttr key share)}\n")
-         (attrNames share)
-    ));
+  settingsFormat = pkgs.formats.ini { };
+  configFile = settingsFormat.generate "smb.conf" cfg.settings;
 
-  configFile = pkgs.writeText "smb.conf"
-    (if cfg.configText != null then cfg.configText else
-    ''
-      [global]
-      security = ${cfg.securityType}
-      passwd program = /run/wrappers/bin/passwd %u
-      invalid users = ${smbToString cfg.invalidUsers}
-
-      ${cfg.extraConfig}
-
-      ${smbToString (map shareConfig (attrNames cfg.shares))}
-    '');
+  samba = cfg.package;
```

Comment:
While we are at it, please drop this very generic variable and replace all occurences with cfg.package